### DISCUSSION TITLE
How to define tabline and winbar?

### DISCUSSION THREAD

**Discussion by unphased** [MunifTanjim/nougat.nvim#40] [Q&A] [answered]
I was able to easily incorporate an example by copying it into my dotfiles and using a config closure in the Lazy entry for Nougat to load it, and statusline looks great. But the bubbly example does not make any changes to my tabline or winbar. Is this intentional? Could we get some simple docs to show what all the settable values are?

**Comment by MunifTanjim** [comment]
Hey, thanks for trying it out.

> But the bubbly example does not make any changes to my tabline or winbar. Is this intentional? Could we get some simple docs to show what all the settable values are?

Using Winbar should be exactly the same as using Statusline. With the following differences:
- Instead of `Bar('statusline')`, do `Bar('winbar')`
- Instead of `bar_util.set_statusline`, use `bar_util.set_winbar`

Tabline support is a bit experimental (well, the whole plugin is experimental 😂). But you can check this for example: https://github.com/MunifTanjim/dotfiles/blob/3d81d787b5/private_dot_config/nvim/lua/plugins/ui/nougat.lua#L272-L305

---

Since this plugin is still work-in-progress, the documentation is a bit lacking. Once I get the time to implement a few more features (e.g. https://github.com/MunifTanjim/nougat.nvim/issues/38), I'll focus on better documentation.

**Reply by MunifTanjim** [reply, marked as answer]
Docs:
- https://github.com/MunifTanjim/nougat.nvim/tree/main/lua/nougat/bar#set_tabline
- https://github.com/MunifTanjim/nougat.nvim/tree/main/lua/nougat/bar#set_winbar

Examples:
- https://github.com/MunifTanjim/nougat.nvim/tree/main/examples

**Reply by unphased** [reply]
awesome thank you. I hope to test your examples soon. 

I have been fiddling with heirline so far (incidentally after first learning about nougat and reading on reddit about people comparing it to heirline...) and it is okay for me, powerful enough for my desired customizations, but... takes quite a bit of time to figure out each time a tweak is desired... could you answer some quick questions about nougat? 

How much control do we have over dynamic width? Can we assign a relative priority to each region (so that it can be used to determine which ones are hidden first based on priority)? Can we override hiding a given region based on available width, even if it could still be visible at that time? control over which regions remain visible in response to window width changes is a very important part of the statusbar for me.

**Reply by unphased** [reply]
I also have some more notes, two of the biggest problems i have with heirline right now are that the following are hard to configure with it as it requires huge additions of conditional logic. 

- I want edited buffers to be highlighted more saliently in bgcolor across statusline and tabline. 
- I want insert mode and command mode and other modes to have customized bgcolor states for similar reasons. 

I believe you have features for the second item here but do you support the first? I have not thought long enough about what to do in the case of combination states (e.g. should i differentiate between being in insert mode on a modified buffer and being in insert mode before having typed anything and the buffer is not unmodified? I think the answer is yes! It would be really awesome to have this.)

**Reply by MunifTanjim** [reply]
> How much control do we have over dynamic width?

There are two type of controls:
- **Breakpoint**: You can render different things based on the total `width` (e.g. https://github.com/MunifTanjim/dotfiles/blob/88ea385806ad4a3abf59e1786e9b00def14dfb42/private_dot_config/nvim/lua/plugins/ui/nougat.lua#L148-L160)
- **Item Priority**: https://github.com/MunifTanjim/nougat.nvim/blob/main/lua/nougat/item/README.md#priority (similar to heirline)

> Can we assign a relative priority to each region (so that it can be used to determine which ones are hidden first based on priority)?

Yes, as described above.

> Can we override hiding a given region based on available width, even if it could still be visible at that time?

I'm not sure I understand the question 🤔

> control over which regions remain visible in response to window width changes is a very important part of the statusbar for me.

`nougat.nvim` fully supports that. You won't miss anything.

---

Just keep in mind that `nougat.nvim` is a bit new, so there can be some bugs here and there, but those are gonna go away with time.

**Reply by MunifTanjim** [reply]
> I want edited buffers to be highlighted more saliently in bgcolor across statusline and tabline.

You mean you want the whole statusline background to be different if buffer is in modified state? Or do you want a specific item to have different background color? 🤔

---

Anyway, you can do both by using the `hl` option for `NougatBar` and `NougatItem`.
- https://github.com/MunifTanjim/nougat.nvim/tree/main/lua/nougat/bar#hl
- https://github.com/MunifTanjim/nougat.nvim/tree/main/lua/nougat/item#hl

Set a function to `hl`, and return the highlight you want. e.g.

```lua
local stl = Bar("statusline", {
  hl = function(_, ctx)
    if vim.bo[ctx.bufnr].modified then
      return "GruvboxRed"
    end
    return "GruvboxBlue"
  end
})

**Reply by unphased** [reply]
Thanks for the response! 

> Breakpoint: You can render different things based on the total width (e.g. https://github.com/MunifTanjim/dotfiles/blob/88ea385806ad4a3abf59e1786e9b00def14dfb42/private_dot_config/nvim/lua/plugins/ui/nougat.lua#L148-L160)

From looking around here it looks like it's about defining a separate function for each width size breakpoint (like buckets in a histogram) which is an acceptable traditional approach to configuring width responsiveness. I seem to recall this is primarily how statusline plugins worked in vim going way back. 

The second scheme for width responsiveness is just to greedily remove low priority (or equivalently insert high priority) components until we run out of width, which is self explanatory. 

> > Can we override hiding a given region based on available width, even if it could still be visible at that time? 
>
> I'm not sure I understand the question 🤔

What I'm asking about is an enhancement to the priority based greedy algorithm where we have another optional property (a width) in addition to the priority. Typically we sort by priority and iterate; if there is still space for the next item, it gets inserted and its size is subtracted from the remaining space. In this enhancement we refrain from adding the element if the width of the window is less than the width value.  

It would be a blend of these two algorithms. I haven't thought a lot yet deal about whether it would be a valuable enhancement but it's an idea that's bounced around in my head a bit. 

I also forgot, another point is around just how granular we are able to make the components. This factors in a lot into the concept of having both width limits and priority assignments. 

As an example to illustrate, in heirline I have configured a few components as follows:

```
local RulerExtraSpace = {
    flexible = 4,
    {
        -- %o = byte offset of cursor in the file
        -- 0x%02B = cursor's byte value shown in hex
        provider = "%o 0x%02B "
    },
    {
        provider = "%o %02B "
    },
    {
        provider = "%o "
    },
    {
        provider = ""
    }
}
```

This is a component of the ruler which progressively shrinks, from showing (at its widest) cursor byte offset and cursor value in hex with `0x` prefix, down to nothing, with some reasonable intermediate contracted variations of that. 

Consider this in concert with 

```
local RulerRestSpace = {
    -- %l = current line number
    -- %L = number of lines in the buffer
    -- %c = column number
    -- %P = percentage through file of displayed window
    flexible = 7,
    {
        provider = "/%L:%c %P "
    }, {
        provider = "/%L:%c "
    }, {
        provider = ":%c "
    }, {
        provider = " "
    }
}
```

and 

```
local FileName = {
    init = function(self)
        self.lfilename = vim.fn.fnamemodify(self.filename, ":.")
        if self.lfilename == "" then self.lfilename = "[No Name]" end
    end,
    hl = { fg = utils.get_highlight("Directory").fg, bold = true },

    flexible = 1000,
    -- TODO decide if this expanding chain is really useful since it does make it copy into a useless path)
    {
        provider = function(self)
            return self.lfilename .. " "
        end,
    },
    {
        provider = function(self)
            return vim.fn.pathshorten(self.lfilename, 4) .. " "
        end,
    },
    {
        provider = function(self)
            return vim.fn.pathshorten(self.lfilename, 3) .. " "
        end,
    },
    {
        provider = function(self)
            return vim.fn.pathshorten(self.lfilename, 2) .. " "
        end,
    },
    {
        provider = function(self)
            return vim.fn.pathshorten(self.lfilename, 1) .. " "
        end,
    }
}
```

As you can see, this is extremely responsive: I have two components of the ruler which each have 4 possible widths to collapse through, and I allow the filename part to try to fill with the full path, the last 4 segments of its path, the last 3, etc. 

The main limitation with this configuration in heirline is that you can see the `flexible` prop which defines priority is applied to the entire collapse-stack! This means that because I chose to prioritize the filename, if i have a long filename, it will clobber and shrink the ruler no matter what. 

I would prefer to be able to assign a priority to EACH of these collapse alternatives (I'm not sure what interface nougat presents for choosing between alternative versions of a component!) so that I can say something like "I would prioritize the visibility of the hex of the char under the cursor over 3 filename path segments but under 2 filename path segments". I would also like to be able to set a min width visibility for a component to add blanket behavior letting me force that component to never appear if width is below that value.

**Reply by MunifTanjim** [reply]
> I also forgot, another point is around just how granular we are able to make the components. This factors in a lot into the concept of having both width limits and priority assignments.


You can go as deep as you want:

```lua
local ruler = Item({
  priority = 1,
  prefix = " ",
  content = {
    Item({
      priority = 3,
      content = "%L",
    }),
    Item({
      priority = 2,
      content = ":%c",
    }),
    Item({
      priority = 1,
      content = " %P",
    }),
  },
  suffix = " ",
}))
```

If you have enough items with more priority than `ruler`, and they fill up most of the spaces, then the remaining spaces will be filled up with the parts of `ruler` according to their priority.

![Kapture 2023-11-20 at 12 17 12](https://github.com/MunifTanjim/nougat.nvim/assets/8050659/17d18fac-b561-4f4f-8245-e38267d871b8)

**Reply by MunifTanjim** [reply]
> In this enhancement we refrain from adding the element if the width of the window is less than the width value.

That is basically hiding element when window size is small enough (regardless of priority), right? You can do that using `hidden`:

```lua
nut.ruler({
  hidden = function(_, ctx)
    return ctx.width < 50
  end,
})
```

^ this item will be hidden whenever window width is less than 50.
